app/controllers/concerns/api_helper.rb

```diff
@@ -11,29 +11,38 @@ module ApiHelper
 
   CHECKSUM_LENGTH_SHA1 = 40
   CHECKSUM_LENGTH_SHA256 = 64
+  CHECKSUM_LENGTH_SHA512 = 128
 
   # Verify checksum
   def verify_checksum
+    secrets = Rails.configuration.x.loadbalancer_secrets
     raise ChecksumError unless params[:checksum].present? &&
                                (params[:checksum].length == CHECKSUM_LENGTH_SHA1 ||
-                                params[:checksum].length == CHECKSUM_LENGTH_SHA256)
+                                params[:checksum].length == CHECKSUM_LENGTH_SHA256 ||
+                                params[:checksum].length == CHECKSUM_LENGTH_SHA512) &&
+                               secrets.any?
 
     # Camel case (ex) get_meetings to getMeetings to match BBB server
     check_string = action_name.camelcase(:lower)
     check_string += request.query_string.gsub(
       /&checksum=#{params[:checksum]}|checksum=#{params[:checksum]}&|checksum=#{params[:checksum]}/, ''
     )
 
-    return if Rails.configuration.x.loadbalancer_secrets.any? do |secret|
-      checksum_sha1 = Digest::SHA1.hexdigest(check_string + secret)
-      checksum_sha256 = Digest::SHA256.hexdigest(check_string + secret)
-      ActiveSupport::SecurityUtils.secure_compare(checksum_sha1, params[:checksum]) ||
-      ActiveSupport::SecurityUtils.secure_compare(checksum_sha256, params[:checksum])
+    checksum_algorithms = Rails.configuration.x.loadbalancer_checksum_algorithms
```

Comment:
Rather than trying all supported checksum algorithms, this should try only the selected one (based on the length of the provided checksum). It still needs a check that the provided checksum matches an enabled algorithm as well, of course.

Reply:
Agree. Although this review should have come before. This is how the application has been dealing with multiple algorithms for more than 1 year. We may revisit this once BBB does what it is expected